valhalla/baldr/graphtile.h

```diff
@@ -588,7 +588,20 @@ class GraphTile : public boost::intrusive_ref_counter<GraphTile, GraphTileRefCou
     if (!invalid_time && (flow_mask & kPredictedFlowMask) && de->has_predicted_speed()) {
       seconds %= midgard::kSecondsPerWeek;
       uint32_t idx = de - directededges_;
-      float speed = predictedspeeds_.speed(idx, seconds);
+      uint32_t bucket = seconds / kSpeedBucketSizeSeconds;
+#ifdef ENABLE_THREAD_SAFE_TILE_REF_COUNT
+      float speed = predictedspeeds_.speed(idx, bucket);
+#else
+      float speed;
+      {
+        BucketSpeed& entry = predicted_speed_cache[idx];
+        if (entry.last_bucket != bucket) {
```

Comment:
i wonder why this is showing the performance improvement. why are we computing the speed for the same edge that often? i mean i know we do it at least a couple times, once during the route, once during recosting for shortcuts removal, once during shape attributes (actually not that last one i guess). i assume those are the culprets??

when i was previously looking into how to speed up the predicted speed look up one of the thoughts i had was vectorizing the decompression algorithm. im not exactly sure if it can be done because the types arent exactly the same but SIMD seemed like an obvious avenue for something like inverse dct.

Reply:
> i wonder why this is showing the performance improvement. why are we computing the speed for the same edge that often? i mean i know we do it at least a couple times, once during the route, once during recosting for shortcuts removal, once during shape attributes (actually not that last one i guess). i assume those are the culprets??

Interesting.. The only thing I can think of is we reset edges if we match a restriction https://github.com/valhalla/valhalla/blob/6901099c3f6aea3b930f2a79bc11df694d80e1e4/valhalla/sif/dynamiccost.h#L381 (if I'm not mistaken the edge can be visited more than once in this case).
But  I'll check my requests one more time. There's is a chance that they were close to each other(in sense of distance) so they used the same tiles

Reply:
i guess during the actual routing we will hit the same edge some number of times and call UpdateLabel so maybe that is where most of the savings are